### DISCUSSION TITLE
Currency Question

### DISCUSSION THREAD

**Discussion by ghost** [evennia/evennia#2410] [Evennia Questions] [answered]
Hello all.
I have an issue with my possible currency system.

A little background:
I'm rather new to Evennia, and haven't done much with it despite having created and deleted about 5 projects now.
I'm working on a MUD based off of Brandon Sanderson's Stormlight Archive books. One of the first things I wanted to do was create a currency system, as currency is one of the biggest parts of that world.
(For those who don't know, each coin basically holds a small amount of "magic", and is discharged and recharged frequently)
There are a total of 15 types of coins.

Two questions here:
1) Is it a really really really bad idea to have every coin as an object? I highly doubt a character would ever have more than a couple hundred, as some coins are worth a thousand times what the base type is. I'd prefer if I could reasonably make all coins individual objects, as it would be way easier to keep track of which coins are "charged" with the magic, but would this be really bad for performance reasons?

2) If having every coin be its own object isn't feasible, how else could I keep track of what coins have the magic intact? Recharging coins and trading for charged ones is a major part of what I have planned.

Thanks in advance for any help!

**Comment by kovitikus** [comment, marked as answer]
I apologize that this post isn't a direct answer to your question. I hope that this information is food for thought and that it helps give you options.

***

The following post references systems related to my ItemHandler. You can find the code for them at the following link, from line 174 and below.

https://github.com/kovitikus/hecate/blob/master/items/item_handler.py#L174

***

I made two custom types of item containers that act as "piles" or "groups" or "stacks" for my objects.

The first one is a quantity container. It takes other objects of the exact same type and adds that object's quantity attribute to the container.

Take for example a single gold coin. It's just a basic Object typeclass, with an attribute on it called wealth. The wealth of this object is equal to 1 gold. Now if I have 2 gold coins I can group them together. The system takes each coin and adds its quantity to the container and deletes the object. Now if I need to split up these coins later on, I can just subtract from the quantity attribute (wealth in this case).

This type of object is great for things that aren't unique, meaning that they can be destroyed and recreated without losing any critical data.

I'll generally try to keep these sorts of items as intangible, meaning that I'll only create physical coins when it's needed for characters to hold them or interact with them physically. Purchasing from merchants or paying other characters can be done simply through attribute addition and subtraction.

***

The second type of object I have is an inventory container. 

This is a container object that has contents within it. Take for example a pile of swords. Perhaps you have 3 different swords in this pile, each with its own unique stats and attributes. In this case, the object is preserved by using `move_to` into the inventory container. 

The inventory container has a name generated that best represents the items within. A variety of items would likely just say "a pile of items", but torches (torches have their own fuel, descriptions, and quality) would be "a pile of torches".

***

I should also note that the different types of stacks are listed here in my `objects.py` module.

https://github.com/kovitikus/hecate/blob/master/items/objects.py#L44

That link is for coin. Further down you will see StackQuantity and StackInventory.

Each of these has their own type of `return_appearance` which will return the quantity of the item as a message to the user or the contents of the stack to the user, respectively.

***

Also, some of the code decisions I made may be tweaked in the future. Such as instead of checking for a typeclass, I may instead check for a tag. By and large though, the concept I put together is exactly what I want and the rest is just minor refactoring.

**Reply by ghost** [reply]
Thanks for the quick reply! I think I can use some form of this, maybe using a coin purse item that creates coin objects on demand, that has an attribute dict or something of "cointype":magicleft for each coin. Still maybe not the best solution, but much better than making thousands of objects.

I definitely think I can make something like this work, thanks!

**Comment by strikaco** [comment]
Interesting problem.

> Is it a really really really bad idea to have every coin as an object?

I would advise against this, yes. 

Backend performance aside (the idmapper is going to struggle to keep all of this in memory) it sounds like hell from a UI perspective. If I want to see what my spell options are and which ones need recharging, how on earth do you intend to *display* that across hundreds of containers?

Maybe you're approaching it too literally. 

* Give a character 15 attributes, each corresponding to a pile of a particular type of coin purse.
* The values stored on each attribute should be a list of integers representing the charges left on individual coins.
* If you want to sell a 20-charge storm coin, pop any 20 you have from char.db.storm_coins.
* If you buy a 5-charge fire coin, add a 5 to char.db.fire_coins.
* You could easily calculate how many fireballs one can throw by sum(char.db.fire_coins)
* You'd have to do appraisals of coin market value on the fly; you can't store that much this way.

**Comment by davewiththenicehat** [comment]
For tracking object charge.

I'm a big fan of Python's date time module.
When you decide on an object type(s).
You could [tag](https://evennia.readthedocs.io/en/latest/Tags.html), the object as currency.
Make a db attribute. Set it to the date time you want the gem to be dun.
`obj.db.infused = datetime.now() + timedelta(seconds=obj.db.infused_time)`

Make a script that gathers all currency. If the infused time is less than the current time, the object is dun.
```python
if obj.db.infused < datetime.now():
    obj.db.infused
```
Unlikely you will need this, just in case.
If you have issues comparing db.infused to datetime.now. Likely datetime did not survive the serialization process. You would want to store the infused time a a string instead. If you need to do this you can convert it back to a date time object with:
`datetime.strptime(infused_time, '%Y-%m-%d %H:%M:%S.%f')`

When a Character attempts to infuse themselves you would want to verify the gems they attempt to use are not past their infused time. Just in case the global script has not been called since some gems have past their infused time.

This way gems go dun on their own and Characters can not use gems that should be dun.

 You are going to have a bear of a time collecting the infused times. Also didn't gems do a secondary affect with soul casters?
Currency is a terribly complex object in Roshar.

**Reply by ghost** [reply]
Oh I wasn’t even thinking about soulcasting. Wow, that would mean I need to keep track of some sort of soulcasting usage attribute so they eventually break. I think I might just use non-sphere gems for that. 

I was thinking about having a global script that removes 1 charge (a diamond chip’s worth) from every sphere every few ingame days but that would be a mess having to go through all the characters and probably not very good performance-wise. (Is it? I don’t really know.) 
It’s probably better to calculate on-demand. (Maybe when a character tries to use it, and when a character enters the room, so that they do continue to go dun on their own) I like the datetime idea, I’ll look into it. 

Roshar has horribly complex mechanics everywhere. This is going to be a challenge, for sure.

**Reply by davewiththenicehat** [reply]
Tags allow you to gather all objects with that type easily.

I think evennia's ticker handler might be what you are looking for. I've never used it. If I remember is calls a callback after a set time.
https://evennia.readthedocs.io/en/latest/TickerHandler.html

**Reply by Griatch** [reply]
For someone not having read the books, the currency system suggested sounds bonkers! :smile:
> I was thinking about having a global script that removes 1 charge (a diamond chip’s worth) from every sphere every few ingame days but that would be a mess having to go through all the characters and probably not very good performance-wise. (Is it? I don’t really know.)

While the TickerHandler could be used for this, consider if you really need it. For example, is it enough to check the charge of your coins when you want to use it (or inspect it)? If so, you can just store when the coin was last charged. When you access the coin you compare the current time with the past time and calculate on the fly what the current charge must be. No overhead, no need to loop over hundreds of coins just to change a number that may or may not actually be used. 

Only time you really need a ticker would be if you wanted, say, a message to pop up (without a user prompting for it) saying "coin X has lost a charge". Which sounds _mildly_ annoying ...

**Reply by ghost** [reply]
Yeah that sounds a bit more than mildly annoying...
I probably don't need it to update them all, I'll just do it on demand.
Thanks for the help!
